### PR TITLE
[GlobalISel][ARM] Legalze set_fpmode and get_fpmode

### PR DESCRIPTION
**Pull request by spavloff**
Implement handling of get/set floating point control modes for ARM in Global Instruction Selector.

### CODE REVIEW

File: llvm/test/CodeGen/ARM/GlobalISel/fpenv.ll [hunk 1]
```diff
@@ -4,6 +4,9 @@
 declare i32 @llvm.get.fpenv.i32()
 declare void @llvm.set.fpenv.i32(i32)
 declare void @llvm.reset.fpenv()
+declare i32 @llvm.get.fpmode.i32()
+declare void @llvm.set.fpmode.i32(i32)
+declare void @llvm.reset.fpmode()
```
Comment (tschuett) [inline llvm/test/CodeGen/ARM/GlobalISel/fpenv.ll:9]: The `declare` s are not needed anymore.

File: llvm/lib/Target/ARM/ARMLegalizerInfo.cpp [hunk 2]
```diff
@@ -159,6 +159,8 @@ ARMLegalizerInfo::ARMLegalizerInfo(const ARMSubtarget &ST) {
 
     getActionDefinitionsBuilder({G_GET_FPENV, G_SET_FPENV}).legalFor({s32});
     getActionDefinitionsBuilder(G_RESET_FPENV).alwaysLegal();
+    getActionDefinitionsBuilder(G_GET_FPMODE).legalFor({s32});
```
Comment (tschuett) [inline llvm/lib/Target/ARM/ARMLegalizerInfo.cpp:162]: Merge with `G_GET_FPENV` ?

File: llvm/lib/Target/ARM/ARMLegalizerInfo.cpp [hunk 3]
```diff
@@ -439,6 +443,23 @@ bool ARMLegalizerInfo::legalizeCustom(LegalizerHelper &Helper, MachineInstr &MI,
...
   }
+  case G_SET_FPMODE: {
+    // New FPSCR = (FPSCR & FPStatusBits) | (Modes & ~FPStatusBits)
+    LLT FPEnvTy = LLT::scalar(32);
+    auto FPEnv = MRI.createGenericVirtualRegister(FPEnvTy);
+    auto Modes = MI.getOperand(0).getReg();
+    MIRBuilder.buildInstr(G_GET_FPENV).addDef({FPEnv});
+    auto StatusBitMask = MIRBuilder.buildConstant(FPEnvTy, ARM::FPStatusBits);
+    auto StatusBits =
+        MIRBuilder.buildAnd(FPEnvTy, FPEnv, StatusBitMask).getReg(0);
```
Comment (arsenm) [inline llvm/lib/Target/ARM/ARMLegalizerInfo.cpp:454]: Drop the .getReg(0)

File: llvm/lib/Target/ARM/ARMLegalizerInfo.cpp [hunk 4]
```diff
@@ -439,6 +443,23 @@ bool ARMLegalizerInfo::legalizeCustom(LegalizerHelper &Helper, MachineInstr &MI,
                              *ConstantInt::get(Ctx, AsInteger));
     break;
   }
+  case G_SET_FPMODE: {
+    // New FPSCR = (FPSCR & FPStatusBits) | (Modes & ~FPStatusBits)
+    LLT FPEnvTy = LLT::scalar(32);
+    auto FPEnv = MRI.createGenericVirtualRegister(FPEnvTy);
+    auto Modes = MI.getOperand(0).getReg();
```
Comment (arsenm) [inline llvm/lib/Target/ARM/ARMLegalizerInfo.cpp:450]: no auto, Register

File: llvm/lib/Target/ARM/ARMLegalizerInfo.cpp [hunk 5]
```diff
@@ -439,6 +443,23 @@ bool ARMLegalizerInfo::legalizeCustom(LegalizerHelper &Helper, MachineInstr &MI,
...
+    MIRBuilder.buildInstr(G_GET_FPENV).addDef({FPEnv});
```
Comment (arsenm) [inline llvm/lib/Target/ARM/ARMLegalizerInfo.cpp:451]: Would be nicer if we had a buildGetFPEnv in MachineIRBuilder

File: llvm/lib/Target/ARM/ARMLegalizerInfo.cpp [hunk 6]
```diff
@@ -439,6 +443,23 @@ bool ARMLegalizerInfo::legalizeCustom(LegalizerHelper &Helper, MachineInstr &MI,
...
+    auto StatusBitMask = MIRBuilder.buildConstant(FPEnvTy, ARM::FPStatusBits);
+    auto StatusBits =
+        MIRBuilder.buildAnd(FPEnvTy, FPEnv, StatusBitMask).getReg(0);
+    auto NotStatusBitMask =
+        MIRBuilder.buildConstant(FPEnvTy, ~ARM::FPStatusBits);
+    auto FPModeBits =
+        MIRBuilder.buildAnd(FPEnvTy, Modes, NotStatusBitMask).getReg(0);
```
Comment (arsenm) [inline llvm/lib/Target/ARM/ARMLegalizerInfo.cpp:458]: Drop the .getReg(0)

File: llvm/lib/Target/ARM/ARMLegalizerInfo.cpp (hunk 5 above)
Comment (tschuett) [reply llvm/lib/Target/ARM/ARMLegalizerInfo.cpp:451]: + 1. Please do not introduce opcodes without extended the builder.

**Review by arsenm — APPROVED** [review]

**Pull request merged**